packages/test/src/test-integration-workflows-with-recorded-logs.ts

```diff
@@ -0,0 +1,397 @@
+    case 'continue-as-new':
+      if (waitAllHandlersFinished) {
+        handlerMayReturn = true;
+        await workflow.condition(workflow.allHandlersFinished);
+      }
+      // If we do not pass waitAllHandlersFinished here then the test occasionally fails. Recall
+      // that this test causes the worker to send a WFT response containing commands
+      // [completeUpdate, CAN]. Usually, that results in the update completing, the caller getting
+      // the update result, and the workflow CANing. However occasionally (~1/30) there is a server
+      // level=ERROR msg="service failures" operation=UpdateWorkflowExecution wf-namespace=default error="unable to locate current workflow execution"
+      // the update caller does not get a response, and the update is included again in the first
+      // WFT sent to the post-CAN workflow run. (This causes the current test to fail unless the
+      // post-CAN run waits for handlers to finish).
+      await workflow.continueAsNew('return', waitAllHandlersFinished);
+      throw new Error('unreachable');
+    case 'failure':
+      throw new workflow.ApplicationFailure('Deliberately failing workflow with an unfinished handler');
+    case 'return':
+      if (waitAllHandlersFinished) {
+        handlerMayReturn = true;
+        await workflow.condition(workflow.allHandlersFinished);
+      }
+      break;
+  }
+}
+
+// These tests confirm that the warning is issued / not issued as appropriate for workflow
+// termination via cancellation, continue-as-new, failure, and return, and that there is no warning
+// when waiting on allHandlersFinished before return or continue-as-new.
+test('unfinished update handler with workflow return', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'update', 'return').testWarningIsIssued(true);
+});
+
+test('unfinished update handler with workflow return waiting for all handlers to finish', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(
+    t,
+    'update',
+    'return',
+    'wait-all-handlers-finished'
+  ).testWarningIsIssued(false);
+});
+
+test('unfinished signal handler with workflow return', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'signal', 'return').testWarningIsIssued(true);
+});
+
+test('unfinished signal handler with workflow return waiting for all handlers to finish', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(
+    t,
+    'signal',
+    'return',
+    'wait-all-handlers-finished'
+  ).testWarningIsIssued(false);
+});
+
+test('unfinished update handler with workflow cancellation', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'update', 'cancellation').testWarningIsIssued(false);
+});
+
+test('unfinished signal handler with workflow cancellation', async (t) => {
+  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'signal', 'cancellation').testWarningIsIssued(false);
```

Comment:
TODO (blocking PR): we are going back to warning on cancellation
```suggestion
  await new UnfinishedHandlersWorkflowTerminationTypeTest(t, 'signal', 'cancellation').testWarningIsIssued(true);
```